crates/evm/core/src/evm.rs

```diff
@@ -46,43 +50,62 @@ pub fn new_evm_with_inspector<'i, 'db, I: InspectorExt + ?Sized>(
         error: Ok(()),
     };
     let spec = ctx.cfg.spec;
-    let precompiles = EthPrecompiles {
-        precompiles: Precompiles::new(PrecompileSpecId::from_spec_id(spec)),
-        spec,
-    }
-    .precompiles;
 
-    FoundryEvm {
+    let mut evm = FoundryEvm {
         inner: RevmEvm::new_with_inspector(
             ctx,
             inspector,
             EthInstructions::default(),
-            PrecompilesMap::from_static(precompiles),
+            get_precompiles(spec),
         ),
-    }
+    };
+
+    inject_precompiles(&mut evm);
+
+    evm
 }
 
 pub fn new_evm_with_existing_context<'a>(
     ctx: EthEvmContext<&'a mut dyn DatabaseExt>,
     inspector: &'a mut dyn InspectorExt,
 ) -> FoundryEvm<'a, &'a mut dyn InspectorExt> {
     let spec = ctx.cfg.spec;
-    let precompiles = EthPrecompiles {
-        precompiles: Precompiles::new(PrecompileSpecId::from_spec_id(spec)),
-        spec,
-    }
-    .precompiles;
 
-    FoundryEvm {
+    let mut evm = FoundryEvm {
         inner: RevmEvm::new_with_inspector(
             ctx,
             inspector,
             EthInstructions::default(),
-            PrecompilesMap::from_static(precompiles),
+            get_precompiles(spec),
         ),
+    };
+
+    inject_precompiles(&mut evm);
+
+    evm
+}
+
+/// Conditionally inject additional precompiles into the EVM context.
+fn inject_precompiles(evm: &mut FoundryEvm<'_, impl InspectorExt>) {
+    if evm.inspector_mut().is_odyssey() {
```

Comment:
accessing by `inspector()` would be preferred here as we don't require mutability here

Reply:
Can be a follow up